keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -0,0 +1,887 @@
+It is worth noting that checking for unknown fields whether via a json unmarshal
+that explicitly breaks if it encounters unknown fields (as in the case of
+create, update, or json patch) or extra logic around a merge step (as in
+strategic merge patch or apply) will be less performant than not doing so.
+[Initial
+benchmarks](https://github.com/kubernetes/kubernetes/pull/104433#issuecomment-901398507) estimate ~20% slower 25-30% more memory consumption.
+
+This is deemed acceptable and insignificant because the use case we are
+replacing is client side validation that happens from kubectl (i.e. via a human
+that should not be too impacted by a slight performance hit). Any existing
+automated clients that would be impacted by performance do not currently have a
+way of leveraging server side validation and thus will remain opted-out of this
+by default (or will be willing to tradeoff the performance if they do choose to
+opt-in to server side schema validation).
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Opt-in API Mechanism
+
+There are a few ways we could allow a client to opt-in to server-side unknown
+field validation, we present two options **Content-Type Header** and **query
+parameter**.
+
+#### Content-Type Header
+
+Requests that contain a kubernetes object, also pass along with it a
+content-type header such as “application/json”. One way to indicate to the
+server that it should use strict schema validation and fail when unknown fields
+are passed is to send a mime-type parameter indicating the strict validation
+such as “application/json;validation=strict”. Alternatively we could use a new
+header such as “X-Kubernetes-Validation:strict”.
+
+One could argue that this is the more appropriate way to parameterize opting-in
+to server side schema validation because on the server we are fundamentally
+treating the content as a different type (“application/json” is json data that
+we don’t care if it has extra fields, “application/json;validation=strict” is
+json data that is sensitive to extra fields).
+
+A precedent for using a mime-type parameter is from how we [receive resources as
+tables](https://kubernetes.io/docs/reference/using-api/_print/#receiving-resources-as-tables).
+
+On the other hand, one could also argue that interpreting input strictly or not
+according to the target schema is independent of the content type and that this
+is an inappropriate way to parameterize validation opt-in.
+
+Another argument against this method is that for patch, strictness is handled
+when decoding the *result* of the patch, so it wouldn’t make sense to use
+Content-Type which is providing information about the inbound request body.
```

Comment:
Although like serialization, different patch types may have different rules about strictness.  Can a strategic merge patch even be strict?

I'm slightly in favor of query parameters, mostly because we can version them by the API version.  Content type has to be versioned with the type (i.e. at some point if we break compat with proto2 we will have to have application/vnd.kubernetes.protobuf.2)

Reply:
Currently, no a strategic merge patch cannot be strict. In theory though, this should be doable. The way I understand SMP to work is that in the `mergeMap` function it goes field-by-field through the existing object and checks to see if data should be merged in from the corresponding field on the patch. If there are any extra/unknown fields in the patch, they will never be encountered and will just be ignored.

The current plan is to modify `mergeMap` to track the fields of the patch as they are merged into the existing object. After all fields that can be merged from the patch are, we will know that any fields on the patch that were not merged are unknown fields and can error. Does this sound reasonable to you?

I’ve added the above along with the additional advantages of the query parameter approach to the doc.

Reply:
Yeah, but it's not called out per type (i.e. you can't describe the parameters of a mime type, just a combo of mime types, so you can document `application/json;v1=blah;strict=true,as=Type` but not just "the `as` parameter to a mime type" since they didn't expect composition).  Parameters + openapi are more discoverable, and while I think this feels a bit closer to mime type, discoverability and documentation win for me.